### PR TITLE
Data Collector Process

### PR DESCRIPTION
**Pull request by harshjain2**
Implemented following features :
Load Data Collectors.
Data Collection Events.
Data Collection Message Sink.
Data Collector Message Logger
Removed x86 version of DataCollector, added AnyCPU Version.

### CODE REVIEW

**Comment by harshjain2** [comment]
@dotnet-bot test this please

File: test/datacollector.UnitTests/DummyDataCollectors.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    [DataCollectorFriendlyName("CustomDataCollector")]
+    [DataCollectorTypeUri("my://custom/datacollector")]
+    public class CustomDataCollector : DataCollector, ITestExecutionEnvironmentSpecifier
+    {
+        public static bool IsInitialized = false;
+        public static bool IsSessionStartedInvoked = false;
+        public static bool IsSessionEndedInvoked = false;
+        public static bool GetTestExecutionEnvironmentVariablesThrowException;
+        public static IEnumerable<KeyValuePair<string, string>> EnvVarList;
```
Comment (codito) [inline test/datacollector.UnitTests/DummyDataCollectors.cs]: Since this is a spy, can we use Moq?

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        [TestInitialize]
+        public void Init()
+        {
+            CustomDataCollector.Reset();
+            DummyMessageSink.Reset();
+
+            this.mockMessageSink = new DummyMessageSink();
+            this.dataCollectionManager = new DataCollectionManager(new DataCollectionAttachmentManager(), new DummyMessageSink());
+
+            this.dataCollectorSettings = string.Format("<DataCollector friendlyName=\"CustomDataCollector\" uri=\"my://custom/datacollector\" assemblyQualifiedName=\"Microsoft.VisualStudio.TestPlatform.DataCollector.UnitTests.CustomDataCollector, datacollector.UnitTests, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a\" codebase=\"{0}\" />", typeof(CustomDataCollector).GetTypeInfo().Assembly.Location);
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >ft.VisualStudio.TestPlatform.DataCollector.UnitTests.CustomDataCollector, datacollector.UnitTests, Version=15.0.0.0, [](start = 173, length = 116)

Derive this from  typeof(CustomDataCollector).GetTypeInfo().Assembly.AssemblyName

File: src/datacollector/DataCollectionManager.cs [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+        /// </summary>
+        internal Dictionary<Type, DataCollectorWrapper> RunDataCollectors { get; private set; }
+
+        /// <inheritdoc/>
+        public IDictionary<string, string> InitializeDataCollectors(string settingsXml)
+        {
+            ValidateArg.NotNull(settingsXml, "settingsXml");
+
+            var sessionId = new SessionId(Guid.NewGuid());
+            var dataCollectionContext = new DataCollectionContext(sessionId);
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: >DataCollectionContext [](start = 44, length = 21)

Please remove DataCollectionContext(TestCase) ctor.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        private DataCollectionManager dataCollectionManager;
+        private string defaultRunSettings = "<?xml version=\"1.0\" encoding=\"utf-16\"?>\r\n<RunSettings>\r\n  <DataCollectionRunSettings>\r\n    <DataCollectors >{0}</DataCollectors>\r\n  </DataCollectionRunSettings>\r\n</RunSettings>";
+        private string dataCollectorSettings;
+        private string dataCollectorSettingsWithWrongUri, dataCollectorSettingsWithoutUri;
+
+        private IMessageSink mockMessageSink;
+
+
+        [TestInitialize]
+        public void Init()
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >Init [](start = 20, length = 4)

See if this can be put in ctor.

File: src/datacollector/DataCollectionManager.cs [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+            this.dataCollectionEnvironmentContext = DataCollectionEnvironmentContext.CreateForLocalEnvironment(dataCollectionContext);
+
+            this.attachmentManager.Initialize(sessionId, sourceDirectory, this.messageSink);
+
+            var executionEnvironmentVariables = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: >executionEnvironmentVariables [](start = 16, length = 29)

Please add a comment with rationale for stringcomparer.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            Assert.ThrowsException<ArgumentNullException>(() =>
+            {
+                this.dataCollectionManager.InitializeDataCollectors(null);
+            });
+        }
+
+        [TestMethod]
+        public void InitializeDataCollectorsShouldReturnEmptyDictionaryIfDataCollectorsAreNotConfigured()
+        {
+            var RunSettings = string.Format(this.defaultRunSettings, string.Empty);
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >RunSettings [](start = 16, length = 11)

nit: camelCase

File: src/datacollector/DataCollectionManager.cs [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+
+        /// <summary>
+        /// Gets cache of data collectors associated with the run.
+        /// </summary>
+        internal Dictionary<Type, DataCollectorWrapper> RunDataCollectors { get; private set; }
+
+        /// <inheritdoc/>
+        public IDictionary<string, string> InitializeDataCollectors(string settingsXml)
+        {
+            ValidateArg.NotNull(settingsXml, "settingsXml");
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: Add tracing

File: src/datacollector/DataCollectionManager.cs [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+            this.dataCollectionEnvironmentContext = DataCollectionEnvironmentContext.CreateForLocalEnvironment(dataCollectionContext);
+
+            this.attachmentManager.Initialize(sessionId, sourceDirectory, this.messageSink);
+
+            var executionEnvironmentVariables = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            var runConfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(settingsXml);
+            sourceDirectory = RunSettingsUtilities.GetTestResultsDirectory(runConfiguration);
+
+            var dataCollectionRunSettings = XmlRunSettingsUtilities.GetDataCollectionRunSettings(settingsXml);
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: XmlRunSettingsUtilities, DataCollectionRunSettings does not have traces. Please add.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        {
+            string runSettings = string.Format(defaultRunSettings, dataCollectorSettings);
+
+            var envVarList = new List<KeyValuePair<string, string>> { new KeyValuePair<string, string>("key", "value") };
+            CustomDataCollector.EnvVarList = envVarList;
+
+            var result = this.dataCollectionManager.InitializeDataCollectors(runSettings);
+
+            Assert.AreEqual(1, result.Count);
+            Assert.AreEqual("key", result.Keys.First());
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >Assert [](start = 12, length = 6)

Use CollectionAssert.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+        {
+            var RunSettings = string.Format(defaultRunSettings, dataCollectorSettings + dataCollectorSettings);
+
+            this.dataCollectionManager.InitializeDataCollectors(RunSettings);
+
+            Assert.AreEqual(1, this.dataCollectionManager.RunDataCollectors.Count);
+            Assert.IsTrue(CustomDataCollector.IsInitialized);
+        }
+
+        [TestMethod]
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >TestMethod [](start = 9, length = 10)

Add test for InitializeShouldNotAddDataCollectorIfItIsDisabled

File: src/datacollector/DataCollectionManager.cs [hunk 11]
```diff
@@ -0,0 +1,665 @@
...
+        /// <summary>
+        /// Loads and initializes datacollector using datacollector settings.
+        /// </summary>
+        /// <param name="dataCollectorSettings">
+        /// The data collector settings.
+        /// </param>
+        private void LoadAndInitialize(DataCollectorSettings dataCollectorSettings)
+        {
+            var collectorTypeName = dataCollectorSettings.AssemblyQualifiedName;
+            var collectorDisplayName = string.IsNullOrWhiteSpace(dataCollectorSettings.FriendlyName) ? collectorTypeName : dataCollectorSettings.FriendlyName;
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: Add a test case for this

File: src/datacollector/DataCollectionManager.cs [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+        /// The assembly.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Type"/>.
+        /// </returns>
+        private static Type GetDataCollectorType(string binaryPath, string dataCollectorTypeName, out Assembly assembly)
+        {
+            assembly = null;
+            Type dctype = null;
+            if (File.Exists(binaryPath))
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: Add test case for negative scenario. May need to use IFileHelper.

File: src/datacollector/DataCollectionManager.cs [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+            {
+                assembly = LoadAssemblyFromPath(binaryPath);
+            }
+
+            dctype = assembly?.GetTypes().FirstOrDefault(type => type.AssemblyQualifiedName != null && type.AssemblyQualifiedName.Equals(dataCollectorTypeName));
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: assembly.GetType(typeName)

File: src/datacollector/DataCollectorConfig.cs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            var typeUri = default(Uri);
+            var typeUriAttributes = GetAttributes(dataCollectorType, typeof(DataCollectorTypeUriAttribute), true);
+
+            var typeUriAttribute = (DataCollectorTypeUriAttribute)typeUriAttributes[0];
+
+            // The type uri can not be null or empty.
+            if (string.IsNullOrEmpty(typeUriAttribute.TypeUri))
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Resources.Resources.DataCollector_TypeIsNull, dataCollectorType.FullName));
```
Comment (codito) [inline src/datacollector/DataCollectorConfig.cs]: >ArgumentException [](start = 26, length = 17)

Need tests for these

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+            var result = this.dataCollectionManager.InitializeDataCollectors(runSettings);
+
+            Assert.AreEqual(0, this.dataCollectionManager.RunDataCollectors.Count);
+            Assert.IsTrue(DummyMessageSink.IsSendMessageInvoked);
+            Assert.IsTrue(CustomDataCollector.IsDisposeInvoked);
+        }
+
+        [TestMethod]
+        public void InitializeDataCollectorsShouldReturnOnlyOneEnvirnmentVariableIfMoreThanOneVariablesWithSameKeyIsSpecified()
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >InitializeDataCollectorsShouldReturnOnlyOneEnvirnmentVariableIfMoreThanOneVariablesWithSameKeyIsSpecified [](start = 20, length = 105)

nit: spell

File: src/datacollector/DataCollectionManager.cs [hunk 16]
```diff
@@ -0,0 +1,665 @@
...
+                this.LogWarning(string.Format(CultureInfo.CurrentCulture, Resources.Resources.DataCollectorAssemblyNotFound, collectorDisplayName));
+                return;
+            }
+            catch (Exception ex)
+            {
+                this.LogWarning(string.Format(CultureInfo.CurrentCulture, Resources.Resources.DataCollectorTypeNotFound, collectorDisplayName, ex.Message));
+                return;
+            }
+
+            if (this.RunDataCollectors.ContainsKey(dataCollectorConfig.DataCollectorType))
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: This check is not required since RunSettings already deduplicates

File: src/datacollector/DataCollectorWrapper.cs [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+    using Microsoft.VisualStudio.TestPlatform.DataCollector.Interfaces;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollector;
+
+    /// <summary>
+    /// Encapsulates datacollector object and other objects required to facilitate datacollection.
+    /// </summary>
+    internal class DataCollectorWrapper
```
Comment (codito) [inline src/datacollector/DataCollectorWrapper.cs]: >DataCollectorWrapper [](start = 19, length = 20)

Rename to DataCollectorInformation

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+            this.dataCollectionManager.InitializeDataCollectors(runSettings);
+            mockDataCollectionAttachmentManager.GetDataThrowException = true;
+
+            var result = this.dataCollectionManager.SessionEnded();
+
+            Assert.IsNull(result);
+        }
+    }
+
+    internal class MockDataCollectionAttachmentManager : IDataCollectionAttachmentManager
```
Comment (codito) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: >MockDataCollectionAttachmentManager [](start = 19, length = 35)

Replace with Moq

File: src/datacollector/DataCollectionManager.cs [hunk 19]
```diff
@@ -0,0 +1,665 @@
...
+        {
+            this.Dispose(true);
+
+            // Use SupressFinalize in case a subclass
+            // of this type implements a finalizer.
+            GC.SuppressFinalize(this);
+        }
+
+        /// <inheritdoc/>
+        public Collection<AttachmentSet> SessionEnded()
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: >SessionEnded [](start = 41, length = 12)

This should not return null. If it does, every consumer has to add an extra null check. May be an empty set if better.

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+        {
+            ValidateArg.NotNull(id, nameof(id));
+
+            ValidateArg.NotNull(messageSink, nameof(messageSink));
+
+            this.messageSink = messageSink;
+
+            if (string.IsNullOrEmpty(outputDirectory))
+            {
+                this.SessionOutputDirectory = Path.Combine(Path.GetTempPath(), DefaultOutputDirectoryName);
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: >SessionOutputDirectory [](start = 21, length = 22)

Path.Combine takes N number of strings

File: src/datacollector/DataCollectionManager.cs [hunk 21]
```diff
@@ -0,0 +1,665 @@
...
+            foreach (var dataCollectorWrapper in this.RunDataCollectors.Values)
+            {
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("DataCollectionManger:SendEvent: Raising event {0} to collector {1}", args.GetType(), dataCollectorWrapper.DataCollectorConfig.FriendlyName);
+                }
+
+                try
+                {
+                    dataCollectorWrapper.Events.RaiseEvent(args);
```
Comment (codito) [inline src/datacollector/DataCollectionManager.cs]: >dataCollectorWrapper [](start = 20, length = 20)

Should be this.events.RaiseEvent

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+
+        /// <summary>
+        /// Gets the session output directory.
+        /// </summary>
+        internal string SessionOutputDirectory { get; private set; }
+
+        /// <summary>
+        /// Gets the list of attachment requests.
+        /// </summary>
+        internal List<AttachmentRequest> AttachmentRequests { get; private set; }
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: >AttachmentRequests [](start = 41, length = 18)

It should be possible to replace this with async Tasks for Copy/Move. We could wait on Task.WhenAll(attachmentRequests)?

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Dispose event object
+        /// </summary>
+        public void Dispose()
+        {
+            // Dispose all attachment requests..
+            this.AttachmentRequests.ForEach(request => request.Dispose());
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: >AttachmentRequests [](start = 17, length = 18)

If AttachmentRequests are not required (in case of Task), we may not need Dispose.

**Comment by codito** [comment]
:clock1:

File: src/datacollector/DataCollectorConfig.cs (hunk 14 above)
Comment (harshjain2) [inline src/datacollector/DataCollectorConfig.cs]: done.

File: src/datacollector/DataCollectorWrapper.cs (hunk 17 above)
Comment (harshjain2) [inline src/datacollector/DataCollectorWrapper.cs]: done.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 6 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: done.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 10 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: done.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 15 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: done.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 18 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: done.

File: test/datacollector.UnitTests/DummyDataCollectors.cs (hunk 1 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DummyDataCollectors.cs]: As we are loading the datacollectors through run settings, it is not possible to dynamically construct the object and load it through run settings.

File: test/datacollector.UnitTests/DataCollectionManagerTests.cs (hunk 2 above)
Comment (harshjain2) [inline test/datacollector.UnitTests/DataCollectionManagerTests.cs]: done.

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 24]
```diff
@@ -196,86 +151,47 @@ public void AddAttachment(FileTransferInformationExtension fileTransferInfo)
             }
         }
 
-        /// <summary>
-        /// Dispose event object
-        /// </summary>
+        /// <inheritdoc/>
         public void Dispose()
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: Why is this required? Do we not allocate a new `DataCollectionAttachmentManager` for data collection session?

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 25]
```diff
@@ -196,86 +151,47 @@ public void AddAttachment(FileTransferInformationExtension fileTransferInfo)
...
 
-            var directoryName = Path.GetDirectoryName(attachmentRequest.LocalFilePath);
+            var directoryName = Path.GetDirectoryName(localFilePath);
+
             if (!Directory.Exists(directoryName))
             {
                 Directory.CreateDirectory(directoryName);
             }
-            else if (File.Exists(attachmentRequest.LocalFilePath))
+            else if (File.Exists(localFilePath))
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: Should deletion of existing file be responsibility of `Validate`?

Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: This should be logged that we're removing existing file. What happens if we cannot remove the file?

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 26]
```diff
@@ -285,141 +201,83 @@ private static void Validate(AttachmentRequest attachmentRequest)
...
-                    attachmentRequest.FileTransferInfo.FileName);
-                CompleteFileTransfer(attachmentRequest, error);
-            }
+            var testCaseId = fileTransferInfo.Context.HasTestCase
+                                 ? fileTransferInfo.Context.TestExecId.Id.ToString()
+                                 : string.Empty;
+
+            var directoryPath = Path.Combine(
+                this.SessionOutputDirectory,
+                testCaseId.ToString());
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: `testCaseId` is already a `string`

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 27]
```diff
@@ -285,141 +201,83 @@ private static void Validate(AttachmentRequest attachmentRequest)
...
+
+            var task = new Task(() =>
+             {
+                 Validate(fileTransferInfo, localFilePath);
+
+                 try
+                 {
+                     if (fileTransferInfo.PerformCleanup)
+                     {
+                         File.Move(fileTransferInfo.FileName, localFilePath);
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: Please log these operations.

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 28]
```diff
@@ -285,141 +201,83 @@ private static void Validate(AttachmentRequest attachmentRequest)
...
+                     {
+                         EqtTrace.Error(
+                             "DataCollectionAttachmentManager.TriggerCallBack: Error occurred while raising the file transfer completed callback for {0}. Error: {1}",
+                             localFilePath,
+                             e.ToString());
+                     }
+                 }
+             });
+
+            this.attachmentTasks.Add(task);
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: We're waiting on the `task`, not the `continuationTask`?

File: src/datacollector/DataCollectionAttachmentManager.cs [hunk 29]
```diff
@@ -285,141 +201,83 @@ private static void Validate(AttachmentRequest attachmentRequest)
...
+                testCaseId.ToString());
+            var localFilePath = Path.Combine(directoryPath, Path.GetFileName(fileTransferInfo.FileName));
+
+            var task = new Task(() =>
+             {
+                 Validate(fileTransferInfo, localFilePath);
+
+                 try
+                 {
+                     if (fileTransferInfo.PerformCleanup)
```
Comment (codito) [inline src/datacollector/DataCollectionAttachmentManager.cs]: It is possible user may want to cancel a test run or the runner/communication channel crash during these operations. What is the strategy in such cases?

File: src/datacollector/DataCollectionAttachmentManager.cs (hunk 20 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionAttachmentManager.cs]: done.

File: src/datacollector/DataCollectionAttachmentManager.cs (hunk 26 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionAttachmentManager.cs]: done.

File: src/datacollector/DataCollectionAttachmentManager.cs (hunk 27 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionAttachmentManager.cs]: done.

File: src/datacollector/DataCollectionAttachmentManager.cs (hunk 28 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionAttachmentManager.cs]: done.

**Comment by harshjain2** [comment]
@dotnet-bot test this please.

**Comment by harshjain2** [comment]
@dotnet-bot test this please.

**Comment by harshjain2** [comment]
@dotnet-bot  test this please

File: src/datacollector/DataCollectionAttachmentManager.cs (hunk 29 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionAttachmentManager.cs]: In order to handle user triggered test run cancellation, we will use cancellation token to cancel the in-progress file transfer tasks. 
For crash, we need to detect the socket communication failure and then cancellation of in-progress transfer operations before exiting.

File: src/datacollector/DataCollectionManager.cs (hunk 19 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionManager.cs]: done.

File: src/datacollector/DataCollectionAttachmentManager.cs (hunk 29 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionAttachmentManager.cs]: Added a todo for now, will handle this scenario with integration of vstest.console process with datacollector process.

**Comment by codito** [comment]
What is the change to fix build issues?

Please undo the change in TestPlatform.targets before merge. It is required for localization targets.

File: src/datacollector/DataCollectionManager.cs (hunk 3 above)
Comment (harshjain2) [inline src/datacollector/DataCollectionManager.cs]: This is getting used in InProcDataCollector. We can revisit this later.
